src/udc.rs

```diff
@@ -0,0 +1,223 @@
+        // Put 0 to a16[0] for offset global state data
+        put     a16[0],0x00                     ;
+        // Extract 128 bits from the beginning of s[1] into r128[1]
+        extr    s16[2],r128[1],a16[0]           ;
+        // a32[0] now has token index from global state
+        spy     a32[0],r128[1]                  ;
+        // Read owned state with index a[0] into s[0]
+        cnv     a8[1],a16[1]                    ;
+        // Read contract state with index a[1] into s[2]
+        lds     0x0FA0,a16[1],s16[2]            ;
+        // Put 0 to a16[0] for offset global state data
+        put     a16[0],0x00                     ;
+        // Extract 128 bits from the beginning of s[1] into r128[1]              
+        extr    s16[2],r128[1],a16[0]           ;
+        // a32[1] now has token index from allocation
+        spy     a32[1],r128[1]                  ;
+        // check that token indexes match
+        eq.n    a32[0],a32[1]                   ;
+        // if they do, jump to the next check
+        jif     0x005E                          ;
+        // if not, inc +1 in contract state count
+        add     a8[1],0x1                       ;
+        jmp     0x0023                          ;
+        // offset_0x0057 (pos 87):
+        // we need to put a string into first string register which will be used as an error
+        // message
+        put     s16[0],"the allocated token has an invalid index";
+        fail                                    ;
+        // offset_0x005E (pos 94):
+        // Put 4 to a16[0]
+        put     a16[0],4                        ;
+        // Extract 128 bits starting from the fifth byte of s[1] into r128[0]
+        extr    s16[2],r128[1],a16[0]           ;
+        // a64[1] now has owned fraction
+        spy     a64[1],r128[1]                  ;
+        // put 1 to a64[0]
+        put     a64[0],1                        ;
+        // check that owned fraction == 1
+        eq.n    a64[0],a64[1]                   ;
+        // if yes, jump to next verification    
+        jif     0x0079                         ;
+        // terminate the subroutine
+        put     s16[0],"owned fraction is not 1";
+        fail                                    ;
+        // offset_0x0079 (pos 121):
+        // do we check all tokens? 
+        eq.n    a8[2],a8[0]                     ;
+        // if yes, jump to next finish program
+        jif     0x0085                          ;
+        // if no, iterate next token index 
+        add     a8[0],0x1                       ;      
+        jmp     0x001F                          ;        
+        // offset_0x0085 (pos 133):
+        // terminate the subroutine
+        ret                                     ;
+
+    };
+
+    let alu_lib = Lib::assemble::<Instr<RgbIsa>>(&code).unwrap();
+    let alu_id = alu_lib.id();
+    const FN_GENESIS_OFFSET: u16 = 0;
+    const FN_TRANSFER_OFFSET: u16 = 0x40;
```

Comment:
Are you sure that this offset points to the correct instruction?

Reply:
No, `FN_TRANSFER_OFFSET` is incorrect. After finish the transfers, I will update these offsets.